internal/datastore/common/relationships.go

```diff
@@ -0,0 +1,181 @@
+	colsToSelect = StaticValueOrAddColumnForSelect(colsToSelect, queryInfo, queryInfo.Schema.ColUsersetRelation, &subjectRelation)
+
+	if !queryInfo.SkipCaveats || queryInfo.Schema.ColumnOptimization == ColumnOptimizationOptionNone {
+		colsToSelect = append(colsToSelect, &caveatName, &caveatCtx)
+	}
+
+	colsToSelect = append(colsToSelect, &expiration)
+
+	if withIntegrity {
+		colsToSelect = append(colsToSelect, &integrityKeyID, &integrityHash, &timestamp)
+	}
+
+	if len(colsToSelect) == 0 {
+		var unused int
+		colsToSelect = append(colsToSelect, &unused)
+	}
+
+	return func(yield func(tuple.Relationship, error) bool) {
+		err := tx.QueryFunc(ctx, func(ctx context.Context, rows R) error {
+			var r Rows = rows
+			if crwe, ok := r.(closeRowsWithError); ok {
+				defer LogOnError(ctx, crwe.Close)
+			} else if cr, ok := r.(closeRows); ok {
+				defer cr.Close()
+			}
+
+			span.AddEvent("Query issued to database")
+			relCount := 0
+			for rows.Next() {
+				if err := rows.Scan(colsToSelect...); err != nil {
+					return fmt.Errorf(errUnableToQueryRels, fmt.Errorf("scan err: %w", err))
+				}
+
+				var caveat *corev1.ContextualizedCaveat
+				if !queryInfo.SkipCaveats || queryInfo.Schema.ColumnOptimization == ColumnOptimizationOptionNone {
+					if caveatName.Valid {
+						var err error
+						caveat, err = ContextualizedCaveatFrom(caveatName.String, caveatCtx)
+						if err != nil {
+							return fmt.Errorf(errUnableToQueryRels, fmt.Errorf("unable to fetch caveat context: %w", err))
+						}
+					}
+				}
+
+				var integrity *corev1.RelationshipIntegrity
+				if integrityKeyID != "" {
+					integrity = &corev1.RelationshipIntegrity{
+						KeyId:    integrityKeyID,
+						Hash:     integrityHash,
+						HashedAt: timestamppb.New(timestamp),
+					}
+				}
+
+				if expiration != nil {
+					// Ensure the expiration is always read in UTC, since some datastores (like CRDB)
+					// will normalize to local time.
+					t := expiration.UTC()
+					expiration = &t
+				}
+
+				relCount++
+				if !yield(tuple.Relationship{
+					RelationshipReference: tuple.RelationshipReference{
+						Resource: tuple.ObjectAndRelation{
+							ObjectType: resourceObjectType,
+							ObjectID:   resourceObjectID,
+							Relation:   resourceRelation,
+						},
+						Subject: tuple.ObjectAndRelation{
+							ObjectType: subjectObjectType,
+							ObjectID:   subjectObjectID,
+							Relation:   subjectRelation,
+						},
+					},
+					OptionalCaveat:     caveat,
+					OptionalExpiration: expiration,
+					OptionalIntegrity:  integrity,
+				}, nil) {
+					return nil
+				}
+			}
+
+			if err := rows.Err(); err != nil {
+				return fmt.Errorf(errUnableToQueryRels, fmt.Errorf("rows err: %w", err))
+			}
+
+			span.AddEvent("Rels loaded", trace.WithAttributes(attribute.Int("relCount", relCount)))
+			return nil
+		}, sqlStatement, args...)
+		if err != nil {
+			if !yield(tuple.Relationship{}, err) {
```

Comment:
if there's an error, shouldn't you unconditionally stop the iterator?

Reply:
Strictly speaking, this would stop the iterator, because the parent QueryRels would have stopped on the error.

Reply:
I think it is clearer to not check the return value of yield and return here instead